### ISSUE TITLE
TextureAtlasManager crash

### ISSUE THREAD

**Issue by Cipkanikolaj** [anegostudios/VintageStory-Issues#4693]
### Game Version

v1.20.0-rc.3

### Platform

Windows

### Modded

Modded

### SP/MP

Singleplayer

### Description

Game crashes, if I understand correctly, because of a texture space? sorry I don't know much about game dev to be of any use.

### How to reproduce

Had similar crash happen periodically, sometimes just running around, opening inventory, alt tabbing etc

### Logs

```
Running on 64 bit Windows 10.0.22631.0 with 16085 MB RAM
Game Version: v1.20.0-rc.3 (Unstable)
12/14/2024 9:45:46 PM: Critical error occurred
Loaded Mods: primitivesurvival@3.7.3, game@1.20.0-rc.3, blocksoverlay@4.0.2, butchering@1.7.3, commonlib@2.6.1, firewoodtosticks@1.0.0, creative@1.20.0-rc.3, survival@1.20.0-rc.3, playercorpse@1.11.0
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
at Vintagestory.Client.NoObf.TextureAtlasManager.AllocateTextureSpace(Int32 width, Int32 height, Int32& textureSubId, TextureAtlasPosition& texPos) in VintagestoryLib\Client\Render\Atlas\TextureAtlasManager.cs:line 390
at Vintagestory.Client.NoObf.TextureAtlasManager.InsertTexture(IBitmap bmp, Int32& textureSubId, TextureAtlasPosition& texPos, Single alphaTest) in VintagestoryLib\Client\Render\Atlas\TextureAtlasManager.cs:line 298
at Vintagestory.Client.NoObf.TextureAtlasManager.GetOrInsertTexture(AssetLocationAndSource loc, Int32& textureSubId, TextureAtlasPosition& texPos, CreateTextureDelegate onCreate, Single alphaTest)
at Vintagestory.GameContent.ShapeTextureSource.get_Item(String textureCode) in VSSurvivalMod\Block\BlockShapeFromAttributes.cs:line 68
at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElement(Int32 indexForLogging, MeshData meshdata, ShapeElement element, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 400
at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 309
at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 309
at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShape(Shape shapeBase, MeshData& modeldata, Vec3f wholeMeshRotation, Vec3f wholeMeshOffset, Single wholeMeshScale, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 250
at Vintagestory.GameContent.BlockShapeFromAttributes.GetOrCreateMesh(IShapeTypeProps cprops, ITexPositionSource overrideTexturesource, String overrideTextureCode) in VSSurvivalMod\Block\BlockShapeFromAttributes.cs:line 653
at Vintagestory.GameContent.BlockShapeFromAttributes.OnBeforeRender(ICoreClientAPI capi, ItemStack itemstack, EnumItemRenderTarget target, ItemRenderInfo& renderinfo) in VSSurvivalMod\Block\BlockShapeFromAttributes.cs:line 477
at Vintagestory.Client.NoObf.InventoryItemRenderer.RenderItemstackToGui(ItemSlot inSlot, Double posX, Double posY, Double posZ, Single size, Int32 color, Single dt, Boolean shading, Boolean origRotate, Boolean showStackSize) in VintagestoryLib\Client\Render\InventoryItemRenderer.cs:line 194
at Vintagestory.API.Client.GuiElementItemSlotGridBase.RenderInteractiveElements(Single deltaTime) in VintagestoryApi\Client\UI\Elements\Impl\Interactive\Inventory\GuiElementItemSlotGridBase.cs:line 514
at Vintagestory.API.Client.GuiComposer.Render(Single deltaTime) in VintagestoryApi\Client\UI\GuiComposer.cs:line 718
at Vintagestory.API.Client.GuiDialog.OnRenderGUI(Single deltaTime) in VintagestoryApi\Client\UI\Dialog\GuiDialog.cs:line 389
at Vintagestory.Client.NoObf.GuiManager.OnRenderFrameGUI(Single deltaTime) in VintagestoryLib\Client\Systems\Gui\GuiManager.cs:line 318
at Vintagestory.Client.NoObf.ClientEventManager.TriggerRenderStage(EnumRenderStage stage, Single dt) in VintagestoryLib\Client\Util\ClientEventManager.cs:line 188
at Vintagestory.Client.NoObf.ClientMain.RenderToDefaultFramebuffer(Single dt) in VintagestoryLib\Client\ClientMain.cs:line 991
at Vintagestory.Client.GuiScreenRunningGame.RenderToDefaultFramebuffer(Single dt) in VintagestoryLib\Client\MainMenu\Screens\GuiScreenRunningGame.cs:line 242
at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 695
at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 670
at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 92
at OpenTK.Windowing.Desktop.GameWindow.Run()
at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs) in VintagestoryLib\Client\ClientProgram.cs:line 334
at Vintagestory.Client.ClientProgram.<>c__DisplayClass10_0.<.ctor>b__1() in VintagestoryLib\Client\ClientProgram.cs:line 130
at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93
```
[client-main.log](https://github.com/user-attachments/files/18137693/client-main.log)
[server-main.log](https://github.com/user-attachments/files/18137694/server-main.log)

**Comment by Craluminum2413** [comment]
Please paste crash as text!

**Comment by Cipkanikolaj** [comment]
Running on 64 bit Windows 10.0.22631.0 with 16085 MB RAM
Game Version: v1.20.0-rc.3 (Unstable)
12/14/2024 9:45:46 PM: Critical error occurred
Loaded Mods: primitivesurvival@3.7.3, game@1.20.0-rc.3, blocksoverlay@4.0.2, butchering@1.7.3, commonlib@2.6.1, firewoodtosticks@1.0.0, creative@1.20.0-rc.3, survival@1.20.0-rc.3, playercorpse@1.11.0
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at Vintagestory.Client.NoObf.TextureAtlasManager.AllocateTextureSpace(Int32 width, Int32 height, Int32& textureSubId, TextureAtlasPosition& texPos) in VintagestoryLib\Client\Render\Atlas\TextureAtlasManager.cs:line 390
   at Vintagestory.Client.NoObf.TextureAtlasManager.InsertTexture(IBitmap bmp, Int32& textureSubId, TextureAtlasPosition& texPos, Single alphaTest) in VintagestoryLib\Client\Render\Atlas\TextureAtlasManager.cs:line 298
   at Vintagestory.Client.NoObf.TextureAtlasManager.GetOrInsertTexture(AssetLocationAndSource loc, Int32& textureSubId, TextureAtlasPosition& texPos, CreateTextureDelegate onCreate, Single alphaTest)
   at Vintagestory.GameContent.ShapeTextureSource.get_Item(String textureCode) in VSSurvivalMod\Block\BlockShapeFromAttributes.cs:line 68
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElement(Int32 indexForLogging, MeshData meshdata, ShapeElement element, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 400
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 309
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShapeElements(MeshData meshdata, ShapeElement[] elements, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 309
   at Vintagestory.Client.NoObf.ShapeTesselator.TesselateShape(Shape shapeBase, MeshData& modeldata, Vec3f wholeMeshRotation, Vec3f wholeMeshOffset, Single wholeMeshScale, TesselationMetaData meta) in VintagestoryLib\Client\Render\Tesselation\ShapeTesselator.cs:line 250
   at Vintagestory.GameContent.BlockShapeFromAttributes.GetOrCreateMesh(IShapeTypeProps cprops, ITexPositionSource overrideTexturesource, String overrideTextureCode) in VSSurvivalMod\Block\BlockShapeFromAttributes.cs:line 653
   at Vintagestory.GameContent.BlockShapeFromAttributes.OnBeforeRender(ICoreClientAPI capi, ItemStack itemstack, EnumItemRenderTarget target, ItemRenderInfo& renderinfo) in VSSurvivalMod\Block\BlockShapeFromAttributes.cs:line 477
   at Vintagestory.Client.NoObf.InventoryItemRenderer.RenderItemstackToGui(ItemSlot inSlot, Double posX, Double posY, Double posZ, Single size, Int32 color, Single dt, Boolean shading, Boolean origRotate, Boolean showStackSize) in VintagestoryLib\Client\Render\InventoryItemRenderer.cs:line 194
   at Vintagestory.API.Client.GuiElementItemSlotGridBase.RenderInteractiveElements(Single deltaTime) in VintagestoryApi\Client\UI\Elements\Impl\Interactive\Inventory\GuiElementItemSlotGridBase.cs:line 514
   at Vintagestory.API.Client.GuiComposer.Render(Single deltaTime) in VintagestoryApi\Client\UI\GuiComposer.cs:line 718
   at Vintagestory.API.Client.GuiDialog.OnRenderGUI(Single deltaTime) in VintagestoryApi\Client\UI\Dialog\GuiDialog.cs:line 389
   at Vintagestory.Client.NoObf.GuiManager.OnRenderFrameGUI(Single deltaTime) in VintagestoryLib\Client\Systems\Gui\GuiManager.cs:line 318
   at Vintagestory.Client.NoObf.ClientEventManager.TriggerRenderStage(EnumRenderStage stage, Single dt) in VintagestoryLib\Client\Util\ClientEventManager.cs:line 188
   at Vintagestory.Client.NoObf.ClientMain.RenderToDefaultFramebuffer(Single dt) in VintagestoryLib\Client\ClientMain.cs:line 991
   at Vintagestory.Client.GuiScreenRunningGame.RenderToDefaultFramebuffer(Single dt) in VintagestoryLib\Client\MainMenu\Screens\GuiScreenRunningGame.cs:line 242
   at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 695
   at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 670
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 92
   at OpenTK.Windowing.Desktop.GameWindow.Run()
   at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs) in VintagestoryLib\Client\ClientProgram.cs:line 334
   at Vintagestory.Client.ClientProgram.<>c__DisplayClass10_0.<.ctor>b__1() in VintagestoryLib\Client\ClientProgram.cs:line 130
   at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93

**Comment by Craluminum2413** [comment]
Try without mods

**Comment by radfast** [comment]
This seems to be a game engine issue, only seen when many/large mods are used.  The game engine is intended to be able to support large mods so I think this seems a valid issue here.  Not yet sure of the cause.

@Cipkanikolaj  please can you post an extract from your `clientsettings.json` file found in the `%appdata%\VintagestoryData` folder.  Don't post the whole file as it contains some personal information like your email address.   But I am interested to see these two settings, found in the `intsettings` section of the file:

    "maxTextureAtlasWidth": xxxx,
    "maxTextureAtlasHeight": xxxx,

**Comment by Craluminum2413** [comment]
Probably related https://github.com/anegostudios/VintageStory-Issues/issues/4723

**Comment by Darkx352** [comment]
I had the same issue. I had to pull the "x" mods down, and the crash would disappear. It didn't matter which mods just reduced their number, and the poof problem was gone. 

Adding this mod completely fixed the problem for me. Other people had an issue with this mod that it would have a default animation setting of 56. and that would kick them. However, once we updated the JSON animations to 230, we had zero issues. 

I did extensive testing on this yesterday, because I thought it was a mod pack causing the issue. So you if have any questions or need me to recreate it. Then just let me know.


[client-crash.log](https://github.com/user-attachments/files/18203180/client-crash.log)
" 
  "intSettings": {
    "fpHandsFoV": 75,
    "mipmapLevel": 3,
    "musicFrequency": 2,
    "graphicsPresetId": 6,
    "webRequestTimeout": 10,
    "maxAnimatedElements": 230,
    "archiveLogFileCount": 5,
    "archiveLogFileMaxSizeMb": 1024,
    "shadowMapQuality": 1,
    "optimizeRamMode": 1,
    "ssaoQuality": 1,
    "viewDistance": 256,
    "fieldOfView": 70,
    "maxFps": 75,
    "particleLevel": 100,
    "maxDynamicLights": 25,
    "weirdMacOSMouseYOffset": 5,
    "screenshotExifDataMode": 0,
    "chunkVerticesUploadRateLimiter": 3,
    "masterSoundLevel": 19,
    "soundLevel": 25,
    "entitySoundLevel": 29,
    "ambientSoundLevel": 33,
    "weatherSoundLevel": 23,
    "musicLevel": 0,
    "screenWidth": 3862,
    "screenHeight": 2182,
    "gameWindowMode": 0,
    "rightDialogMargin": 0,
    "leftDialogMargin": 0,
    "maxQuadParticles": 8000,
    "maxCubeParticles": 4000,
    "maxAsyncQuadParticles": 80000,
    "maxAsyncCubeParticles": 80000,
    "itemCollectMode": 0,
    "mouseSmoothing": 30,
    "mouseSensivity": 50,
    "modelDataPoolMaxVertexSize": 500000,
    "modelDataPoolMaxIndexSize": 750000,
    "modelDataPoolMaxParts": 1500,
    "maxTextureAtlasWidth": 4096,
    "maxTextureAtlasHeight": 2048,
    "godRays": 0,
    "chatWindowWidth": 700,
    "chatWindowHeight": 200,
    "recordingBufferSize": 60,
    "vsyncMode": 0,
    "windowBorder": 0,
    "guiColorsPreset": 1,
    "schematicMaxUploadSizeKb": 200,
    "playerVoiceLevel": 33,
    "traderVoiceLevel": 32
  },
  "boolSettings": {
    "selectedBlockOutline": true,
    "showMoreGfxOptions": false,
    "dynamicColorGrading": true,
    "multipleInstances": false,
    "showSurvivalHelpDialog": true,
    "showCreativeHelpDialog": true,
    "smoothShadows": true,
    "flipScreenshot": true,
    "renderClouds": true,
    "renderParticles": true,
    "transparentRenderPass": true,
    "extendedDebugInfo": false,
    "showentitydebuginfo": false,
    "showBlockInfoHud": true,
    "showBlockInteractionHelp": true,
    "showCoordinateHud": false,
    "wavingStuff": true,
    "renderMetaBlocks": false,
    "ambientParticles": true,
    "pauseGameOnLostFocus": false,
    "viewBobbing": true,
    "invertMouseYAxis": false,
    "highQualityAnimations": true,
    "occlusionculling": true,
    "developerMode": false,
    "glDebugMode": false,
    "showWhitelistedServers": true,
    "showModdedServers": true,
    "showPasswordProtectedServers": true,
    "showOpenForAllServers": true,
    "liquidFoamAndShinyEffect": true,
    "testGlExtensions": true,
    "offThreadMipMaps": false,
    "toggleSprint": true,
    "separateCtrlKeyForMouse": false,
    "allowSettingHRTFaudio": true,
    "useHRTFaudio": false,
    "force48khzHRTFaudio": true,
    "fxaa": true,
    "autoChat": true,
    "bloom": true,
    "skipNvidiaProfileCheck": false,
    "scaleScreenshot": false,
    "immersiveMouseMode": false,
    "startupErrorDialog": false,
    "hasGameServer": false,
    "showMinimapHud": false,
    "chatdialogvisible": true
  },
  "floatSettings": {
    "guiScale": 1.5,
    "fontSize": 1.0,
    "lodBias": 0.33,
    "lodBiasFar": 0.67,
    "blockAtlasSubPixelPadding": 0.01,
    "itemAtlasSubPixelPadding": 0.0,
    "gammaLevel": 2.2,
    "extraGammaLevel": 1.0,
    "brightnessLevel": 1.0,
    "sepiaLevel": 0.21755669,
    "cameraShakeStrength": 1.0,
    "wireframethickness": 1.0,
    "previewTransparency": 0.3,
    "fpHandsYOffset": 0.0,
    "swimmingMouseSmoothing": 0.9,
    "mouseWheelSensivity": 1.0,
    "ambientBloomLevel": 0.0,
    "extraContrastLevel": 0.10241392,
    "recordingFrameRate": 30.0,
    "gameTickFrameRate": -1.0,
    "instabilityWavingStrength": 1.0,
    "ssaa": 0.5,
    "megaScreenshotSizeMul": 2.0,
    "minbrightness": 0.0"

**Comment by radfast** [comment]
Thanks, that is helpful, so in summary, the crash triggered with many mods and default max texture size settings:

    "maxTextureAtlasWidth": 4096,
    "maxTextureAtlasHeight": 2048,

That's a concern, we need to fix the core game code.  While AutoAtlas was certainly helpful as a temporary fix (at least once it updated maxAnimatedElements to 230 ;) ) it should not be necessary, these days the game is now intended to support unlimited numbers of textures (as long as your GPU has enough RAM) so people can have as many mods as they like.  And in theory even the 8192 x 8192 which AutoAtlas sets - 8 times larger than vanilla in total - might not be enough if someone has _all the mods_

**Comment by radfast** [comment]
Fixed in upcoming 1.20.0-rc.6.  Thanks for the reports and info, very helpful
